app/main.py

```diff
@@ -1,214 +1,96 @@
-            mordred["power"] += mordred["potion"]["effect"]["power"]
-
-        if "protection" in mordred["potion"]["effect"]:
-            mordred["protection"] += mordred["potion"]["effect"]["protection"]
-
-        if "hp" in mordred["potion"]["effect"]:
-            mordred["hp"] += mordred["potion"]["effect"]["hp"]
-
-    # red_knight
-    red_knight = knightsConfig["red_knight"]
-
-    # apply armour
-    red_knight["protection"] = 0
-    for a in red_knight["armour"]:
-        red_knight["protection"] += a["protection"]
-
-    # apply weapon
-    red_knight["power"] += red_knight["weapon"]["power"]
-
-    # apply potion if exist
-    if red_knight["potion"] is not None:
-        if "power" in red_knight["potion"]["effect"]:
-            red_knight["power"] += red_knight["potion"]["effect"]["power"]
-
-        if "protection" in red_knight["potion"]["effect"]:
-            red_knight["protection"] += red_knight["potion"]["effect"]["protection"]
-
-        if "hp" in red_knight["potion"]["effect"]:
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
-
-    # -------------------------------------------------------------------------------
-    # BATTLE:
-
-    # 1 Lancelot vs Mordred:
-    lancelot["hp"] -= mordred["power"] - lancelot["protection"]
-    mordred["hp"] -= lancelot["power"] - mordred["protection"]
-
-    # check if someone fell in battle
-    if lancelot["hp"] <= 0:
-        lancelot["hp"] = 0
-
-    if mordred["hp"] <= 0:
-        mordred["hp"] = 0
-
-    # 2 Arthur vs Red Knight:
-    arthur["hp"] -= red_knight["power"] - arthur["protection"]
-    red_knight["hp"] -= arthur["power"] - red_knight["protection"]
-
-    # check if someone fell in battle
-    if arthur["hp"] <= 0:
-        arthur["hp"] = 0
-
-    if red_knight["hp"] <= 0:
-        red_knight["hp"] = 0
+        potion=Potion(
+            name=knightsconfig["mordred"]["potion"]["name"],
+            effect=knightsconfig["mordred"]["potion"]["effect"],
+        )
+        if knightsconfig["mordred"]["potion"]
+        else None,
+    )
+
+    red_knight = Knight(
+        name="Red Knight",
+        power=knightsconfig["red_knight"]["power"],
+        hp=knightsconfig["red_knight"]["hp"],
+        weapon=Weapon(
+            name=knightsconfig["red_knight"]["weapon"]["name"],
+            power=knightsconfig["red_knight"]["weapon"]["power"],
+        ),
+        armour=[
+            Armour(part=a["part"], protection=a["protection"])
+            for a in knightsconfig["red_knight"]["armour"]
+        ],
+        potion=Potion(
+            name=knightsconfig["red_knight"]["potion"]["name"],
+            effect=knightsconfig["red_knight"]["potion"]["effect"],
+        )
+        if knightsconfig["red_knight"]["potion"]
+        else None,
+    )
+
+    lancelot.take_damage(mordred.total_power)
+    mordred.take_damage(lancelot.total_power)
+    arthur.take_damage(red_knight.total_power)
+    red_knight.take_damage(arthur.total_power)
```

Comment:
The damage application between knights is handled directly here. It would be more appropriate to use the `Battle` class to manage the fight logic, ensuring consistency and encapsulation of battle-related operations. Consider refactoring to instantiate `Battle` objects and call their `fight` method instead.